tests/Microsoft.Identity.Test.Unit.net45/PublicApiTests/AccountTests.cs

```diff
@@ -56,5 +66,129 @@ public void Constructor_PropertiesSet()
+                var pcaCn = InitPcaForCloud(AzureCloudInstance.AzureChina, httpManager, TokenCacheFile);
+
+                // Act
+                var accountsGlobal = await pcaGlobal.GetAccountsAsync().ConfigureAwait(false);
+                var accountsDe = await pcaDe.GetAccountsAsync().ConfigureAwait(false);
+                var accountsCn = await pcaCn.GetAccountsAsync().ConfigureAwait(false);
+
+                // Assert
+                Assert.AreEqual("login.microsoftonline.com", accountsGlobal.Single().Environment);
+                Assert.AreEqual("login.microsoftonline.de", accountsDe.Single().Environment);
+                Assert.AreEqual("login.chinacloudapi.cn", accountsCn.Single().Environment);
+            }
+        }
+
+        // Bug https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/1030
+        [TestMethod]
+        [DeploymentItem(@"Resources\MultiCloudTokenCache.json")]
+        public async Task GermanCloudNoNetworkCallAsync()
+        {
+            // if a network call is made, this test will fail
+            IMsalHttpClientFactory factoryThatThrows = Substitute.For<IMsalHttpClientFactory>();
+            factoryThatThrows.When(x => x.GetHttpClient()).Do(x => { Assert.Fail("A network call is being performed"); });
+
+            var pcaDe = PublicClientApplicationBuilder
+              .Create("0615b6ca-88d4-4884-8729-b178178f7c27")
+              .WithAuthority(AzureCloudInstance.AzureGermany, AadAuthorityAudience.PersonalMicrosoftAccount)
+              .WithHttpClientFactory(factoryThatThrows)
+              .BuildConcrete();
+
+            pcaDe.InitializeTokenCacheFromFile(ResourceHelper.GetTestResourceRelativePath("MultiCloudTokenCache.json"));
+
+            // remove all but the German account
+            pcaDe.UserTokenCacheInternal.Accessor.GetAllAccounts()
+                .Where(a => a.Environment != "login.microsoftonline.de")
+                .ToList()
+                .ForEach(a => pcaDe.UserTokenCacheInternal.Accessor.DeleteAccount(a.GetKey()));
+
+            // Act
+            var accountsDe = await pcaDe.GetAccountsAsync().ConfigureAwait(false);
+
+            // Assert
+            Assert.AreEqual("login.microsoftonline.de", accountsDe.Single().Environment);
+        }
+
+
+        private PublicClientApplication InitPcaForCloud(AzureCloudInstance cloud, HttpManager httpManager, string tokenCacheFile)
+        {
+            PublicClientApplication pca = PublicClientApplicationBuilder
+                  .Create("0615b6ca-88d4-4884-8729-b178178f7c27")
+                  .WithAuthority(cloud, AadAuthorityAudience.PersonalMicrosoftAccount)
+                  .WithHttpManager(httpManager)
+                  .BuildConcrete();
+
+            pca.InitializeTokenCacheFromFile(ResourceHelper.GetTestResourceRelativePath(tokenCacheFile));
+            pca.UserTokenCacheInternal.Accessor.AssertItemCount(3, 3, 3, 3, 1);
+
+            return pca;
+        }
     }
```

Comment:
what about b2c users and/or b2c users in different clouds?

Reply:
If I understand correctly there is no instance discovery in b2c and in this case our code will use the env passed in to do all the checks, i.e. preferred_network == preferred_cache. I will add a unit test.